File: Android-Native/app/src/test/java/com/kkunan/randomdogandroid/features/randomimage/data/repositories/RandomDogRepositoryImplTest.kt
package com.kkunan.randomdogandroid.features.randomimage.data.repositories

import com.kkunan.randomdogandroid.common.domain.responses.LocalResponse
import com.kkunan.randomdogandroid.features.randomimage.data.datasources.RandomDogNetworkDatasource
import com.kkunan.randomdogandroid.features.randomimage.data.models.RandomDogImageResponseModel
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.junit.Assert.*

import org.junit.After
import org.junit.Before
import org.junit.Test
import kotlin.random.Random

@ExperimentalCoroutinesApi
class RandomDogRepositoryImplTest {

    lateinit var repository: RandomDogRepositoryImpl
    lateinit var datasource: RandomDogNetworkDatasource

    @Before
    fun setUp() {
        datasource = mockk()
        repository = RandomDogRepositoryImpl(datasource)
    }

    @Test
    fun randomImage_should_pass_correct_parameters_to_datasource() {
        runTest {
            // arrange
            coEvery { datasource.getRandomImage(any()) } answers {
                RandomDogImageResponseModel(null, "failed")
            }
            val expected = Random.nextInt()

            // act
            repository.randomImage(expected)

            // assert
            coVerify { datasource.getRandomImage(expected) }
        }
    }

    @Test
    fun randomImage_should_return_failed_if_status_is_failed(){
        runTest {
            // arrange
            coEvery { datasource.getRandomImage(any()) } answers {
                RandomDogImageResponseModel(null, "failed")
            }

            // act
            val response = repository.randomImage(2)

            // assert
            if (response is LocalResponse.FAILURE){
                assertEquals(response.error, "Cannot get dog images :'(")
            } else fail()
        }
    }

    @Test
    fun randomImage_should_return_correct_urls_if_status_is_success(){
        runTest {
            // arrange
            val urls = listOf("url1", "url2")

            coEvery { datasource.getRandomImage(any()) } answers {
                RandomDogImageResponseModel(urls, "success")
            }

            // act
            val response = repository.randomImage(2)

            // assert
            if (response is LocalResponse.SUCCESS){
                assertEquals(response.value.map { it.url }, urls)
            } else fail()
        }
    }
}

File: random_dog_flutter/test/features/randomdog/data/repositories/random_dog_repository_impl_test.dart
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'package:random_dog_flutter/common/data/responses/local_response.dart';
import 'package:random_dog_flutter/features/randomdog/data/datasources/network/random_image_network_datasource.dart';
import 'package:random_dog_flutter/features/randomdog/data/models/random_dog_response.dart';
import 'package:random_dog_flutter/features/randomdog/data/repositories/random_dog_repository_impl.dart';
import 'package:random_dog_flutter/features/randomdog/domain/entities/dog_image.dart';

import 'random_dog_repository_impl_test.mocks.dart';

@GenerateMocks([RandomImageNetworkDatasource])
void main(){
  late MockRandomImageNetworkDatasource datasource;
  late RandomDogRepositoryImpl repository;

  setUp((){
    datasource = MockRandomImageNetworkDatasource();
    repository = RandomDogRepositoryImpl(datasource);
  });

  test('randomImage_should_pass_correct_parameters_to_datasource', () async {
    // arrange
    when(datasource.randomImage(any)).thenAnswer((realInvocation) async {
      return RandomDogResponse(message: [], status: 'status');
    });
    final number = Random().nextInt(1000);

    // act
    await repository.randomImage(number);

    // assert
    verify(datasource.randomImage(number));
  });

  test('randomImage_should_return_error_if_datasource_message_null', () async {
    // arrange
    when(datasource.randomImage(any)).thenAnswer((realInvocation) async {
      return RandomDogResponse(message: null, status: 'status');
    });
    const expected = LocalResponse<DogImage>(value: null, error: "Cannot get image!");

    // act
    final response = await repository.randomImage(4);

    // assert
    expect(response, expected);
  });

  test('randomImage_should_map_each_url_to_dogImage_and_return_success', () async {
    // arrange
    final expected = DogImage(imageUrls: ['url1', 'url2']);
    when(datasource.randomImage(any)).thenAnswer((realInvocation) async {
      return RandomDogResponse(message: expected.imageUrls, status: 'status');
    });

    // act
    final response = await repository.randomImage(4);

    // assert
    expect(response, LocalResponse(value: expected));

  });
}